### ISSUE TITLE
Exceptions in ReactiveSectionedRecyclerAdapter

### ISSUE THREAD

**Issue by amulware** [toggl/mobileapp#3295]
## 💥 Crash Report
1.4 introduced several new exceptions in the `ReactiveSectionedRecyclerAdapter`. I'm creating only one issue because my hope is that there is a fix for all these probably related problems.

These issues doubled our crash numbers, so fixing them in imperative.

### 👥 Impact
- 70 crashes (almost half of all crashes in 1.4)
- all entirely new in 1.4

### 🛰 Link
https://appcenter.ms/orgs/Toggl/apps/Toggl-Android/crashes/groups/13922b61fdd73dfa1071a7b71e6d04448b50712c
https://appcenter.ms/orgs/Toggl/apps/Toggl-Android/crashes/groups/4199e05e789bc34d6b34775689082b55837576ba
https://appcenter.ms/orgs/Toggl/apps/Toggl-Android/crashes/groups/6d22004065fbad231d1ee6ba4ace2a25ba991940

### 📝  Exception & Stack Trace
`System.InvalidOperationException: Collection was modified; enumeration operation may not execute.`
```cs
List`1+Enumerator[T].MoveNext ()
ReactiveSectionedRecyclerAdapter`3[TModel,TItemViewHolder,TSectionViewHolder].updateSectionIndexes ()
ReactiveSectionedRecyclerAdapter`3[TModel,TItemViewHolder,TSectionViewHolder].UpdateChanges (System.Collections.Generic.IReadOnlyCollection`1[T] changes)
AnonymousSafeObserver`1[T].OnNext (T value)
ObserveOn`1+Context+_[TSource].OnNextPosted (System.Object value)
SyncContext+<>c__DisplayClass2_0.<Post>b__0 ()
Thread+RunnableImplementor.Run ()
IRunnableInvoker.n_Run (System.IntPtr jnienv, System.IntPtr native__this)
(wrapper dynamic-method) System.Object.34(intptr,intptr)
```

`System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. Parameter name: index`
```cs
Requires.Range (System.Boolean condition, System.String parameterName, System.String message)
ImmutableList`1+Node[T].ItemRef (System.Int32 index)
ImmutableList`1[T].get_Item (System.Int32 index)
ReactiveSectionedRecyclerAdapter`3[TModel,TItemViewHolder,TSectionViewHolder].removeRow (Toggl.Foundation.MvvmCross.Collections.Changes.RemoveRowCollectionChange removeRow)
ReactiveSectionedRecyclerAdapter`3[TModel,TItemViewHolder,TSectionViewHolder].UpdateChanges (System.Collections.Generic.IReadOnlyCollection`1[T] changes)
AnonymousSafeObserver`1[T].OnNext (T value)
ObserveOn`1+Context+_[TSource].OnNextPosted (System.Object value)
SyncContext+<>c__DisplayClass2_0.<Post>b__0 ()
Thread+RunnableImplementor.Run ()
IRunnableInvoker.n_Run (System.IntPtr jnienv, System.IntPtr native__this)
(wrapper dynamic-method) System.Object.34(intptr,intptr)
```

`System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. Parameter name: index`
```cs
List`1[T].get_Item (System.Int32 index)
ReactiveSectionedRecyclerAdapter`3[TModel,TItemViewHolder,TSectionViewHolder].GetItemFromAdapterPosition (System.Int32 position)
ReactiveSectionedRecyclerAdapter`3[TModel,TItemViewHolder,TSectionViewHolder].OnBindViewHolder (Android.Support.V7.Widget.RecyclerView+ViewHolder holder, System.Int32 position)
RecyclerView+Adapter.n_OnBindViewHolder_Landroid_support_v7_widget_RecyclerView_ViewHolder_I (System.IntPtr jnienv, System.IntPtr native__this, System.IntPtr native_holder, System.Int32 position)
(wrapper dynamic-method) System.Object.47(intptr,intptr,intptr,int)
```

**Comment by daividssilverio** [comment]
Most likely to be the same source of errors from #3288. Working on it.

**Comment by daividssilverio** [comment]
This is fixed in sodium. Close this when #3286 gets merged.

**Comment by daividssilverio** [comment]
Actually, still related to the same issue, this is still here. And it's related to underlying `ObservableGroupedOrderedCollection`. We might need to add some locking around our enumerators.
```
List`1+Enumerator[T].MoveNext ()
Enumerable+SelectEnumerableIterator`2[TSource,TResult].MoveNext ()
System.Collections.Generic.List`1[T]..ctor (System.Collections.Generic.IEnumerable`1[T] collection) <0x7f8358ddf8 + 0x0015f> in :0
ReactiveSectionedRecyclerAdapter`3[TModel,TItemViewHolder,TSectionViewHolder].flattenItems (Toggl.Foundation.MvvmCross.Collections.ObservableGroupedOrderedCollection`1[TItem] groupsSource)
ReactiveSectionedRecyclerAdapter`3+<>c__DisplayClass17_0[TModel,TItemViewHolder,TSectionViewHolder].<startCurrentCollectionUpdate>b__0 ()
Thread+RunnableImplementor.Run ()
IRunnableInvoker.n_Run (System.IntPtr jnienv, System.IntPtr native__this)
(wrapper dynamic-method) System.Object.34(intptr,intptr)
```

**Comment by amulware** [comment]
Still a thing in 1.6: https://appcenter.ms/orgs/Toggl/apps/Toggl-Android/crashes/errors/2433215786u/overview

Accounts for 5% of all crashes.

**Comment by daividssilverio** [comment]
I didn't check this for a while. I'll get back to it later if no one does.

**Comment by daividssilverio** [comment]
I tried to mess up the ReactiveSectionedRecyclerAdapter so bad but I couldn't replicate the bug. I tried continuing and deleting time entries in a background thread a lot while running sync.
Tried changing the collection a lot near the place where the crash happens... Tried adding a bunch of items near the crash as well...
I wonder what is the state of the app when people are getting this error :0

I'd like someone else to to try replicate this as well. I Can't do it.

**Comment by heytherewill** [comment]
I also could not reproduce this one. That's a pretty tricky one. Maybe even worth adding the `mvexit` label and put on hold?

**Comment by juanlaube** [comment]
Still a thing in 1.9.

**Comment by juanlaube** [comment]
I cannot reproduce these either.

**Comment by dshr** [comment]
![image](https://user-images.githubusercontent.com/6031512/50484047-87bafd80-09ff-11e9-8fbb-55f229811380.png)

Tried reproducing under shoddy network conditions, but no luck in reproducing this.

**Comment by simonrozsival** [comment]
If you can't beat them, ~join~ track them! 😄

**Comment by daividssilverio** [comment]
This is the latest crash log: https://appcenter.ms/orgs/Toggl/apps/Toggl-Android/crashes/errors/2197408927u/overview

We can add a lock in the enumerator returned from `Toggl.Foundation.MvvmCross.Collections.ObservableGroupedOrderedCollection` to avoid enumerations when the collection is getting modified, that could fix this.

**Comment by daividssilverio** [comment]
There's one extra nasty thing we could do to avoid the enumerations when the collection is being modified. We could add a small delay before starting the update process, such that only the last `ICollectionChange` of a batch of updates would trigger the ui update process, which enumerates the time entries collection.
